src/components/data-exploration/differential-expression-tool/DiffExprResults.jsx

```diff
@@ -21,75 +21,38 @@ import AdvancedFilteringModal from 'components/data-exploration/differential-exp
 
-const { Text } = Typography;
-
 const DiffExprResults = (props) => {
   const {
     experimentId, onGoBack, width, height,
   } = props;
 
   const dispatch = useDispatch();
-  const loading = useSelector((state) => state.differentialExpression.properties.loading);
-  const data = useSelector((state) => state.differentialExpression.properties.data);
-  const total = useSelector((state) => state.differentialExpression.properties.total);
-  const error = useSelector((state) => state.differentialExpression.properties.error);
-  const comparisonGroup = useSelector((state) => state.differentialExpression.comparison.group);
-  const comparisonType = useSelector((state) => state.differentialExpression.comparison.type);
-  const advancedFilters = useSelector((state) => (
-    state.differentialExpression.comparison.advancedFilters));
+  const {
+    data, total, error, loading,
+  } = useSelector((state) => state.differentialExpression.properties);
+  const {
+    type: comparisonType,
+    group: comparisonGroup,
+    advancedFilters,
+  } = useSelector((state) => state.differentialExpression.comparison);
   const { properties } = useSelector(getCellSets());
 
   const [dataShown, setDataShown] = useState(data);
   const [exportAlert, setExportAlert] = useState(false);
   const [settingsListed, setSettingsListed] = useState(false);
   const [advancedFilteringModalVisible, setAdvancedFilteringModalVisible] = useState(false);
   const [pathwayAnalysisModalVisible, setPathwayAnalysisModalVisible] = useState(false);
+  const [columns, setColumns] = useState([]);
   const geneTableState = useRef({});
-  const columns = [
-    {
-      title: 'logFC',
-      key: 'logFC',
-      sorter: true,
-      showSorterTooltip: false,
-    },
-    {
-      title: 'adj p-value',
-      key: 'p_val_adj',
-      sorter: true,
-      showSorterTooltip: false,
-      render: (score, record) => <Tooltip title={`adj p-value: ${record.p_val_adj}`}>{score}</Tooltip>,
-    },
-    {
-      title: 'Pct 1',
-      key: 'pct_1',
-      sorter: true,
-      showSorterTooltip: {
-        title: 'The percentage of cells where the feature is detected in the first group',
-      },
-    },
-    {
-      title: 'Pct 2',
-      key: 'pct_2',
-      sorter: true,
-      showSorterTooltip: {
-        title: 'The percentage of cells where the feature is detected in the second group',
-      },
-    },
-    {
-      title: 'AUC',
-      key: 'auc',
-      sorter: true,
-      showSorterTooltip: {
-        title: 'Area under the ROC curve',
-      },
-    },
-  ];
 
   // When data changes, update rows.
+  const buildColumns = (rowData) => columnDefinitions
+    .filter(({ key }) => rowData[0][key] !== undefined);
+
   useEffect(() => {
-    if (data && properties) {
-      setDataShown(data);
-    }
+    if (!data.length || !Object.keys(properties).length) return null;
+    setDataShown(data);
+    setColumns(buildColumns(data));
```

Comment:
maybe its better to set columns first and then set data, makes more sense other than that.

Reply:
switched